src/mrpro/algorithms/optimizers/pgd.py

```diff
@@ -0,0 +1,158 @@
+                    max_iterations=200,
+                    backtrack_factor=1.0,
+                )
+
+
+    Parameters
+    ----------
+    f
+        convex, differentiable functional
+    g
+        convex, non-smooth functional with computable proximal map
+    initial_value
+        initial value for the solution of the algorithm
+    stepsize
+        stepsize needed in the gradient step, constant if no backtracking is used, otherwise it is the initial stepsize
+    reg_parameter
+        regularization parameter that multiplies g
+    max_iterations
+        number of iterations
+    backtrack_factor
+        must be :math:`<=1`. if :math:`<1.`, backtracking rule for stepsize following https://www.ceremade.dauphine.fr/~carlier/FISTA
+        is used
+    callback
+        function to be called at each iteration
+
+    Returns
+    -------
+        an approximate solution of the minimization problem
+    """
+    # check that backtracking factor is in the correct range
+    if not 0.0 <= backtrack_factor <= 1.0:
+        raise ValueError('Backtracking factor must be in the range [0, 1].')
+
+    backtracking = not math.isclose(backtrack_factor, 1)
+
+    if isinstance(initial_value, torch.Tensor):
+        initial_values: tuple[torch.Tensor, ...] = (initial_value,)
+    else:
+        initial_values = initial_value
+
+    if isinstance(g, ProximableFunctional):
+        g_sum = ProximableFunctionalSeparableSum(g)
+    else:
+        g_sum = g
+
+    x_old = initial_values
+    y: tuple[torch.Tensor, ...] = initial_values
+    gradient: tuple[torch.Tensor, ...]
+    f_y: torch.Tensor
+    t_old = 1.0
+    grad_and_value_f = torch.func.grad_and_value(
+        lambda x: f(*x)[0],
+    )
+    for iteration in range(max_iterations):
+        while stepsize > 1e-30:
+            # calculate the proximal gradient step
+            gradient, f_y = grad_and_value_f(y)
+            x = g_sum.prox(
+                *[yi - stepsize * gi for yi, gi in zip(y, gradient, strict=True)], sigma=reg_parameter * stepsize
+            )
+
+            if not backtracking:
+                # no need to check stepsize, continue to next iteration
+                break
+            difference = tuple(xi - yi for xi, yi in zip(x, y, strict=True))
+            quadratic_approx = (
+                f_y
+                + 1 / (2 * stepsize) * sum(di.abs().square().sum() for di in difference)
+                + sum(torch.vdot(gi.flatten(), di.flatten()).real for gi, di in zip(gradient, difference, strict=True))
+            )
+
+            (f_x,) = f(*x)
+            if f_x <= quadratic_approx:
+                # stepsize is ok, continue to next iteration
+                break
+            stepsize *= backtrack_factor
+
+        else:
+            if backtracking:
+                raise RuntimeError('After backtracking, the stepsize became to small.')
+            else:
+                raise RuntimeError('Stepsize to small.')
+
+        # update timestep t
```

Comment:
is this a "timestep" ?

Reply:
what do you mean? throughout the code I always called it stepsize